src/mame/drivers/magicard.cpp

```diff
@@ -350,313 +571,656 @@ TODO: check this register,doesn't seem to be 100% correct.
-					bitmap.pix(y, (x * 4) + 3) = m_palette->pen(color);
-
-				color = ((m_magicram[count]) & 0x00f0) >> 4;
-
-				if(cliprect.contains((x * 4) + 2, y))
-					bitmap.pix(y, (x * 4) + 2) = m_palette->pen(color);
-
-				color = ((m_magicram[count]) & 0x0f00) >> 8;
-
-				if(cliprect.contains((x * 4) + 1, y))
-					bitmap.pix(y, (x * 4) + 1) = m_palette->pen(color);
-
-				color = ((m_magicram[count]) & 0xf000) >> 12;
-
-				if(cliprect.contains((x * 4) + 0, y))
-					bitmap.pix(y, (x * 4) + 0) = m_palette->pen(color);
-
-				count++;
-			}
-		}
-	}
-	else  // 8bpp gfx
+	src = m_scc66470->line(scanline);
+	for(int x = 0 ; x < 768 ; x++)
 	{
-		for(int y = 0; y < 300; y++)
-		{
-			for(int x = 0; x < 168; x++)
-			{
-				uint32_t color;
-
-				color = ((m_magicram[count]) & 0x00ff) >> 0;
-
-				if(cliprect.contains((x * 2) + 1, y))
-					bitmap.pix(y, (x * 2) + 1) = m_palette->pen(color);
-
-				color = ((m_magicram[count]) & 0xff00) >> 8;
-
-				if(cliprect.contains((x * 2) + 0, y))
-					bitmap.pix(y, (x * 2) + 0) = m_palette->pen(color);
-
-				count++;
-			}
-		}
+		*dest++ = m_palette->pen(*src++);
 	}
-
 	return 0;
 }
 
+/*************************
+*      R/W Handlers      *
+*************************/
 
-/*********************************************
-*                R/W Handlers                *
-*********************************************/
+uint8_t magicard_state::read_ds1207(offs_t offset)
+{
+	return m_ds1207->read_dq() ? 0x08 : 0x00;
+}
 
-uint16_t magicard_state::test_r()
+void magicard_state::write_ds1207(offs_t offset, uint8_t data)
 {
-	return machine().rand();
+	m_ds1207->write_rst(data&1);
+	m_ds1207->write_clk((data>>1)&1);
+	m_ds1207->write_dq((data>>3)&1);
 }
 
-uint16_t magicard_state::philips_66470_r(offs_t offset)
+void magicard_base_state::output_w(offs_t offset, uint16_t data)
 {
-	switch(offset)
-	{
-		case 0/2:
-		{
-			uint8_t vdisp;
-			vdisp = m_screen->vpos() < 256;
-
-			// TODO: other bits
-			return (m_pcab_vregs[offset] & 0xff7f) | vdisp << 7;
-		}
-	}
+	// bit  0 - counter out
+	// bit  1 - ??
+	// bit  2 - ??
+	// bit  3 - counter jackpot
+	// bit  4 - counter checkout
+	// bit  5 - ??
+	// bit  6 - ??
+	// bit  7 - hold 3 lamp
+	// bit  8 - start lamp
+	// bit  9 - hold 1 lamp
+	// bit 10 - hold 5 lamp
+	// bit 11 - hold 2 lamp
+	// bit 12 - hold 4 lamp
+	// bit 13 - clear lamp
+	// bit 14 - hopper drive
+	// bit 15 - counter in
+}
+
 
-	return m_pcab_vregs[offset];
+uint8_t hotslots_state::read_ds1207_ds2401(offs_t offset)
+{
+	m_ds2401->write( true );
+	return (m_ds2401->read() ? 0x10 : 0x00) | (m_ds1207->read_dq() ? 0x08 : 0x00) ;
 }
 
-void magicard_state::philips_66470_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+void hotslots_state::write_ds1207_ds2401(offs_t offset, uint8_t data)
 {
-	COMBINE_DATA(&m_pcab_vregs[offset]);
+	m_ds2401->write( data & 0x10 ? true : false );
+	m_ds1207->write_rst(data&1);
+	m_ds1207->write_clk((data>>1)&1);
+	m_ds1207->write_dq((data>>3)&1);
```

Comment:
More of the same – mixed spacing styles and things that are less error-prone to use the `BIT` helper for.